Engineering/Frontend/Webassembly.md

```diff
@@ -0,0 +1,47 @@
+---
+tags: engineering/frontend, webassembly, performance
+author: Tran Tien An
+date: 2022-12-24
+---
+
+### What
+
+WebAssembly (abbreviated _Wasm_) was launched in 2017 as a low-level assembly-like language with a compact binary format, so it’s fast to load, execute, and run with near-native performance.
+It is designed as a portable target for the compilation of high-level languages like C/C++/Rust, enabling deployment on the web for client and server applications. That means you do not write WebAssembly, you compile other high-level languages to it.
+
+![[wasm-architecture.png]]
+
+### How
+
+With JavaScript, the code is included in the website and is interpreted as it runs because JavaScript variables are dynamic. The only way to know what the types are for sure is to monitor the code as it executes, which is what the JavaScript engine does. Once the engine is satisfied that it knows the variable’s types, it can convert that section of code into machine code.
+
+![[wasm-js-how-it-work.png]]
+
+WebAssembly isn’t interpreted but, rather, is compiled into the WebAssembly binary format by a developer ahead of time. Because the variable types are all known ahead of time, when the browser loads the WebAssembly file, the JavaScript engine doesn’t need to monitor the code. It can simply compile the code’s binary format into machine code.
+
+![[wasm-how-it-works.png]]
+
+### High-Level Goals
+
+- Be executed at near-native speeds.
+- Be readable and easily debuggable.
+- Be secure.
+- Don’t break the web.
```

Comment:
It would be clearer if we could provide more context or examples for the bullet points.